### PR TITLE
Fixes #18311 - Incorrect quick info for ValueTuple<T>

### LINKED ISSUE #15508: IDE extract method fails to handle some deconstruction scenarios.
The IDE code is not resilient against arbitrary expressions appearing on the left-hand-side of a deconstruction. It tries to produce a zero-element tuple type, which causes an exception to be thrown.

This will be a problem after generalized deconstruction support is added to the compiler, even in error scenarios. I'll skip the two affected tests, `GenerateMethodTests.TestDeconstruction3` and `4`. The offending code is here, in `src\Workspaces\CSharp\Portable\LanguageServices\CSharpTypeInferenceService.TypeInferrer.cs`:

```cs
            private bool TryGetTupleTypesAndNames(
                SeparatedSyntaxList<ArgumentSyntax> arguments,
                out ImmutableArray<ITypeSymbol> elementTypes,
                out ImmutableArray<string> elementNames)
            {
                elementTypes = default(ImmutableArray<ITypeSymbol>);
                elementNames = default(ImmutableArray<string>);

                var elementTypesBuilder = ArrayBuilder<ITypeSymbol>.GetInstance();
                var elementNamesBuilder = ArrayBuilder<string>.GetInstance();
                try
                {
                    foreach (var arg in arguments)
                    {
                        var expr = arg.Expression;
                        if (expr.IsKind(SyntaxKind.DeclarationExpression))
                        {
                            AddTypeAndName((DeclarationExpressionSyntax)expr, elementTypesBuilder, elementNamesBuilder);
                        }
                        else if (expr.IsKind(SyntaxKind.TupleExpression))
                        {
                            AddTypeAndName((TupleExpressionSyntax)expr, elementTypesBuilder, elementNamesBuilder);
                        }
                    }

                    if (elementTypesBuilder.Contains(null))
                    {
                        return false;
                    }

                    elementTypes = elementTypesBuilder.ToImmutable();
                    elementNames = elementNamesBuilder.ToImmutable();
                    return true;
                }
                finally
                {
                    elementTypesBuilder.Free();
                    elementNamesBuilder.Free();
                }
            }
```

there appears to be an `else` missing when we test `expr.IsKind`. This causes the caller to attempt to produce a tuple type with an insufficient number of elements.

### LINKED ISSUE #18311: Incorrect quick info for ValueTuple<T>
**Version Used**: VS 2017 (15.0.0+26228.9)

**Steps to Reproduce**:

1. Use the following code:

    ```csharp
    void Foo()
    {
        var y = ValueTuple.Create(1);
    }
    ```

2. Hover over `var`
3. Hover over `y`

**Expected Behavior**:

* In step 2, the result should be `struct System.ValueTuple<T>`, with `T is int`.
* In step 3, the result should be `ValueTuple<int, int> y`

**Actual Behavior**:

* In step 2, the result is `(System.Int32)`
* In step 3, the result is `(int) y`

Neither of these are valid tuple syntaxes in C#.

### PR DESCRIPTION
**Pull request by dopare**
Fixes #18311 
Closes #15508

Hovering `var` will show `System.ValueTuple<System.Int32>`
Hovering `y` will show `(local variable) ValueTuple<int> y`

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Need tests :) 

They can be added to: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs

Also, tagging @dotnet/roslyn-compiler  as this changes compiler code.

**Comment by dopare** [comment]
Fixing tests

**Comment by CyrusNajmabadi** [comment]
Also, does the same issue for VB? If so, we should fix that as well.

**Comment by dopare** [comment]
VB is ok

**Comment by jcouv** [comment]
If there is any chance this same method is invoked again on the "rest" of the tuple (which nests the elements after the 7th), I'd recommend testing the case of an 8-tuple: `(1, 2, ..., 8)` or `ValueTuple.Create(1, 2, ..., 7, ValueTuple.Create(8))`.

**Comment by dopare** [comment]
There are 2 tests for long tuples here https://github.com/dopare/roslyn/blob/f7c0a5a260c0011e6817db9d0d4d8204d951e1b2/src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs#L285-L353

**Comment by jcouv** [comment]
Nice. LGTM and thanks!

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 1]
```diff
@@ -381,6 +381,11 @@ private bool CanUseTupleTypeName(INamedTypeSymbol tupleSymbol)
         {
             INamedTypeSymbol currentUnderlying = tupleSymbol.TupleUnderlyingType;
 
+            if (currentUnderlying.Arity == 1)
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: Does this need to be <= 1?  Is there a similar issue with zero arity value tuples?

**Comment by CyrusNajmabadi** [comment]
Still need quick info tests.  Thanks! :)

**Comment by dopare** [comment]
Need help. Cannot make unit test to work properly.
`var y$$ = ValueTuple.Create(1);` generates `(local variable) ? y` as actual description

**Comment by CyrusNajmabadi** [comment]
Do you have the right usings in place in the test?

**Comment by CyrusNajmabadi** [comment]
Note: look at TestResources.NetFX.ValueTuple.tuplelib_cs as well.  It appears as if some tests manually concat that in to get the definition of ValueTuple, since it's not part of mscorlib.

**Comment by dopare** [comment]
I am using 

```
await TestInMethodAsync(
@"
var y$$ = ValueTuple.Create(1);
",
                MainDescription("(local variable) ValueTuple<int> y"));
```
`TestInMethodAsync` imports `System`

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs (hunk 1 above)
Comment (dopare) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: For < 1 case `AddNameAndTypeArgumentsOrParameters` method is not called

**Comment by jcouv** [comment]
You'll need to manually include the source for ValueTuple into the test. The mscorlib used in tests doesn't have ValueTuple yet.
Cyrus pointed out the resource string that will help.
Or you can just inline a small implementation ([example](https://github.com/dotnet/roslyn/pull/10994/files#diff-c3f63358589cdd259670f76fa5274e33R1446)).

**Comment by dopare** [comment]
Thanks could fix the issue. Working on new tests.

**Comment by CyrusNajmabadi** [comment]
Your "Adding new tests" commit doesn't seem to be adding any new tests :)

**Comment by dopare** [comment]
Sorry, committed wrong file. Fixed.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review of the compiler change.

**Comment by dopare** [comment]
Squashed commits for next review

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs [hunk 2]
```diff
@@ -307,9 +307,9 @@ static void M()
   .maxstack  2
   .locals init ((int, int, int, int, int, int, int, int) V_0) //x
   IL_0000:  ldloc.0
-  IL_0001:  ldfld      ""int System.ValueTuple<int, int, int, int, int, int, int, (int)>.Item4""
+  IL_0001:  ldfld      ""int System.ValueTuple<int, int, int, int, int, int, int, ValueTuple<int>>.Item4""
   IL_0006:  ldloc.0
-  IL_0007:  ldfld      ""(int) System.ValueTuple<int, int, int, int, int, int, int, (int)>.Rest""
+  IL_0007:  ldfld      ""ValueTuple<int> System.ValueTuple<int, int, int, int, int, int, int, ValueTuple<int>>.Rest""
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: The corresponding VB test should have required the same change.

**Comment by cston** [comment]
Consider adding a simple `SymbolDisplay` test for both C# and for VB. See `SymbolDisplayTests.Tuple`.

**Comment by dopare** [comment]
Updated VB logic and tests. Working on new symbol display tests.

**Comment by sharwell** [comment]
> Hovering `var` will show `System.ValueTuple<System.Int32>`

I believe the conclusion in #18298 was we want to display `int` instead of `System.Int32` for this case. However, it's possible that the work to implement this behavior will be part of the other pull request.

**Comment by sharwell** [comment]
@dopare It's possible that this pull request fixed the following issue which I found during an investigation of #15508. If that is the case, then this pull request likely fixes the last remaining issue with that issue.

```csharp
void Foo()
{
    ValueTuple<int> y = ValueTuple.Create(1);
    y.Item1.ToString(); // Extract method for this line crashes the refactoring provider
}
```

**Comment by dopare** [comment]
Added simple symbol display tests for C# and VB.
@sharwell Will check it.

**Comment by dopare** [comment]
@sharwell I have manually tested this in latest Visual Studio 2017 with and without my changes.

Without my changes `ExtractMethodCodeRefactoringProvider` keeps crashing.

With my changes I could extract following method:

```
void Foo()
{
      ValueTuple<int> y = ValueTuple.Create(1);
      NewMethod(y);
}

private static void NewMethod(ValueTuple<int> y)
{
      y.Item1.ToString(); // Extract method for this line crashes the refactoring provider
}
```

**Comment by cston** [comment]
LGTM thanks.

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs (hunk 2 above)
Comment (sharwell) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: ➡️ As an update, this is now corrected.

**Comment by dopare** [comment]
Squashed all commits

**Comment by sharwell** [comment]
@dopare The Extract Method tests are in [ExtractMethodTests.cs](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs) and [ExtractMethodTests.vb](https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/VisualBasicTest/ExtractMethod/ExtractMethodTests.vb). I'm planning to add tests for the case I mentioned after this PR is merged, but if you want to tackle them now that you already fixed the issue feel free to. 😄

**Comment by dopare** [comment]
@sharwell Will add those tests.

**Comment by dopare** [comment]
@sharwell Added new extract method test for C# and VB.

**Comment by sharwell** [comment]
Look at @dopare, doing my work for me. 😎

**Comment by dopare** [comment]
Wow thanks 👍

**Pull request merged**